ELK components for user-tracking analytics

lbjay opened the pull request:
OK, I think this has settled enough to try again.

Includes:

  * install/configure Elasticsearch, Logstash & Kibana
  * nginx reverse proxy for exposing kibana & ES Kopf plugin
  * Matterhorn user action harvester
  * refactor of ssl cert generation

djcp commented on libraries/default.rb:
```diff
@@ -229,6 +229,69 @@ def git_repo_url(git_data)
+    end
+
+    def create_ssl_cert(ssl_info)
+     
```
Remove this newline, please! I generally do not have spacing after "def" or before the end of a function.

djcp commented on libraries/default.rb:
```diff
@@ -229,6 +229,69 @@ def git_repo_url(git_data)
+        end
+      end 
+
+      ! cert_content.empty?
```
I don't think this is necessary here, it was used in the original to modify how the `engage-nginx-proxy-conf.erb` template rendered if the ssl cert was specifically not defined (e.g. removing the dummy cert).

djcp commented on recipes/configure-engage-nginx-proxy.rb:
```diff
@@ -17,44 +17,8 @@
-
-certificate_exists = ! cert_content.empty?
+ssl_info = node.fetch(:ssl, get_dummy_cert)
+certificate_exists = create_ssl_cert(ssl_info)
```
oh. I see. So the issue is the "create_ssl_cert" function now has two responsibilities - creating the cert and telling you if the cert exists or not. I think I'd rather have a "cert_defined?(ssl_info)" or "cert_exists?(ssl_info)" function for that so that the responsibility is clearer.

lbjay replied:
The logic necessary for this is awkward, and it makes me wonder do we even need it? I'm not clear under what situation I'd want to not have the cert at all, i.e., not even the dummy cert.

lbjay replied:
I tried another stab at it anyway.

djcp commented on libraries/default.rb:
```diff
@@ -428,17 +491,22 @@ def initialize_database(current_deploy_root)
-    def install_init_scripts(current_deploy_root, matterhorn_repo_root)
-      log_dir = node.fetch(:matterhorn_log_directory, '/var/log/matterhorn')
-
+    def xmx_ram_for_this_node(xmx_ram_ratio)
```
OK. I generally wouldn't refactor code if you don't need to.

djcp replied:
ahh, hard to read. You're extracting stuff and I like that.

lbjay replied:
Agreed. The diff is really hard to read, but I think the end result makes sense.

djcp commented on recipes/install-elasticsearch.rb:
```diff
@@ -0,0 +1,105 @@
+  })
+end
+
+directory "/vol/data" do
```
so naming the volume "/vol/data" makes it sound like you could use it for stuff other than elasticsearch, and this recursive chown right here would mess this up. Could we call this "/vol/elasticsearch_data" or perhaps put elasticsearch stuff under "/vol/data/elasticsearch" ?  I bring it up only because of the chef "deploy chown surprise" I ran into recently.

lbjay replied:
Good point. I also realized i'm hardcoding the path there when it should be `data_path`.

djcp commented on recipes/install-elasticsearch.rb:
```diff
@@ -0,0 +1,105 @@
+end
+
+include_recipe "mh-opsworks-recipes::update-package-repo"
+install_package("elasticsearch=#{es_version}")
```
Are any relevant elasticsearch logs getting rotated by default?

lbjay replied:
Elasticsearch handles it's own logrotation by default. The config is [here](https://github.com/elastic/elasticsearch/blob/148265bd164cd5a614cd020fb480d5974f523d81/distribution/src/main/resources/config/logging.yml) which now that I look closely at it doesn't look very good. It's rolling them over but not compressing and (unless i'm missing it) does not have any retention/delete settings. Annoyingly it looks like to do this properly I'll have to deploy a custom logging.yml **and** a logrotate config.

lbjay replied:
I think I'm going to go with this approach: http://stackoverflow.com/questions/28696069/elasticsearch-how-to-configure-logging-yml

djcp replied:
The second approach on that page seems to imply you can get compression without logrotate but I'm not entirely sure.

lbjay replied:
I don't want to rely on log4j for this. Plus for consistency sake. See 605a3cb81aee78234c7a2b89cf4cfeb76d0a8fe2.

djcp commented on recipes/install-elasticsearch.rb:
```diff
@@ -0,0 +1,105 @@
+    cluster_name: es_cluster_name,
+    data_path: data_path
+  })
+  notifies :restart, "service[elasticsearch]", :immediately
```
this is a pattern I need to use better.

djcp commented on recipes/install-elasticsearch.rb:
```diff
@@ -0,0 +1,105 @@
+  mode '644'
+end
+
+http_request "delete_templates" do
```
it looks like these two HTTP requests mean that this recipe is not idempotent - correct?

lbjay replied:
Hmm, not in my thinking. Run it multiple times and you'll always get the same result: the `dce-useractions` template (and any future `dce-*` templates) will be defined. The delete request was me wanting a way to clear out deprectated index templates.

djcp replied:
so the index data will be OK even if the templates get replaced on every boot?

lbjay replied:
The existing index data will not be affected by having the mapping deleted and replaced. The mapping is applied to incoming documents at index time.

I'm taking this out anyway as I realize my intentions here were half-baked. Any modification of existing templates needs to be done much more carefully.

djcp commented on recipes/install-kibana.rb:
```diff
@@ -0,0 +1,59 @@
+
+kibana_version = elk_info[:kibana_version]
+kibana_checksum = elk_info[:kibana_checksum]
+download_path = "#{::Chef::Config[:file_cache_path]}/kibana.tar.gz" 
```
FWIW, I'm going to use this pattern in another couple recipes to avoid a coupling to s3 I've got.

djcp commented on recipes/install-ua-harvester.rb:
```diff
@@ -0,0 +1,77 @@
+stack_name = stack_shortname
+sqs_queue_name = "#{stack_name}-user-actions"
+region = node["opsworks"]["instance"]["region"]
+loggly_info = node.fetch(:loggly, { token: '', url: '' })
```
maybe it's worth extracting this repetition into a function?

lbjay replied:
you're talking about the just the loggly_info (also used in the rsyslog setup), right?

djcp replied:
yeah, and maybe the loggly config.

djcp commented on recipes/install-ua-harvester.rb:
```diff
@@ -0,0 +1,77 @@
+                  ''
+                end
+
+install_package('python-pip')
```
I'd probably include the repo updater recipe at the top of this file.

lbjay replied:
oops, yep.

djcp commented on templates/default/elasticsearch.yml.erb:
```diff
@@ -0,0 +1,13 @@
+cluster.name: <%= @cluster_name %>
+path.data: <%= @data_path %>
+node.name: ${HOSTNAME}
+network.host: 0.0.0.0
```
elasticsearch runs on a private instance, right?

lbjay replied:
yes but I'm going to change it to bind to the private ip anyway

djcp commented on templates/default/elk-nginx-proxy-conf.erb:
```diff
@@ -0,0 +1,59 @@
+  }
+  
+  location ~ ^/es.*$ {
+    proxy_pass http://<%= @elasticsearch_host %>:9200;
```
does this mean elasticsearch is open broadly to public hosts because of this proxy rule?  Maybe we can talk about access / authorization.

lbjay replied:
the `auth_basic` settings defined above this cover all of these locations. There are more sophisticated access/auth mechanisms I plan to explore (e.g., [Elasticsearch Shield](https://www.elastic.co/products/shield)), but for now I'm hoping basic auth + forcing https is enough.

djcp replied:
Cool, the auth coverage wasn't apparent to me. Basic auth + https is definitely good enough.

lbjay commented:
Refactored to support running everything on a single "analytics" instance. Also added automated elasticsearch snapshotting which was on the todo list.

djcp commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 ## TO BE RELEASED
 
+* *OPTIONAL EDITS TO THE CLUSTER CONFIG*:
+  Recipes, etc for setting up ELK. See README.elk.md in mh-opsworks for setup
```
README.analytics.md ?

lbjay commented:
Ugh. This PR should have been against develop.
